File: LCD1602.swift
public struct LCD1602 {
    let clearDisplay: UInt8 = 0x01
    let returnHome: UInt8 = 0x02
    let entryModeSet: UInt8 = 0x04
    let displayControl: UInt8 = 0x08
    let cursorShift: UInt8 = 0x10
    let functionSet: UInt8 = 0x20
    let setCRamAddr: UInt8 = 0x40
    let setDRamAddr: UInt8 = 0x80

    let entryShiftIncrement: UInt8 = 0x01
    let entryLeft: UInt8 = 0x02

    let blinkOn: UInt8 = 0x01
    let cursorOn: UInt8 = 0x02
    let displayOn: UInt8 = 0x04

    let moveRight: UInt8 = 0x04
    let displayMove: UInt8 = 0x08

    let set5x10Dots: UInt8 = 0x04
    let set2Line: UInt8 = 0x08
    let set8BitMode: UInt8 = 0x10

    let backlight: UInt8 = 0x08
    let enableBit: UInt8 = 0x04

    let address: UInt8 = 0x27

    let character: UInt8 = 1
    let command: UInt8 = 0

    let maxLines: UInt8 = 2
    let maxChars: UInt8 = 16

    func writeByte(value: UInt8) {
        var val = value
        i2c_write_blocking(get_default_i2c(), address, &val, 1, false)
    }

    func toggleEnable(value: UInt8) {
        let delay: UInt64 = 600
        sleep_us(delay)
        writeByte(value: value | enableBit)
        sleep_us(delay)
        writeByte(value: value & ~enableBit)
        sleep_us(delay)
    }

    func sendByte(value: UInt8, mode: UInt8) {
        let high: UInt8 = mode | (value & 0xF0) | backlight
        let low: UInt8 = mode | ((value << 4) & 0xF0) | backlight

        writeByte(value: high)
        toggleEnable(value: high)
        writeByte(value: low)
        toggleEnable(value: low)
    }

    func clear() {
        sendByte(value: clearDisplay, mode: command)
    }

    func setCursor(line: UInt8, position: UInt8) {
        let value: UInt8 = (line == 0) ? 0x80 + position : 0xC0 + position;
        sendByte(value: value, mode: command)
    }

    func writeChar(_ val: UInt8) {
    sendByte(value: val, mode: character)
}

func writeStaticString(_ string: StaticString) {
    for i in 0..<string.utf8CodeUnitCount {
        let byte = string.utf8Start[i]
        writeChar(byte)
    }
}

    init() {
        
        i2c_init(get_default_i2c(), 100_000)
        gpio_set_function(UInt32(PICO_DEFAULT_I2C_SDA_PIN), GPIO_FUNC_I2C)
        gpio_set_function(UInt32(PICO_DEFAULT_I2C_SCL_PIN), GPIO_FUNC_I2C)
        gpio_pull_up(UInt32(PICO_DEFAULT_I2C_SDA_PIN))
        gpio_pull_up(UInt32(PICO_DEFAULT_I2C_SCL_PIN))

        sendByte(value: 0x03, mode: command)
        sendByte(value: 0x03, mode: command)
        sendByte(value: 0x03, mode: command)
        sendByte(value: 0x02, mode: command)

        sendByte(value: entryModeSet | entryLeft, mode: command)
        sendByte(value: functionSet | set2Line, mode: command)
        sendByte(value: displayControl | displayOn | ~cursorOn | ~blinkOn, mode: command)


        clear()
        
    }
}
    



  

File: LCD1602.c
#include "LCD1602.h"
#include "hardware/i2c.h"
//#include "pico/binary_info.h"

// commands
static const int clear_display =    0b0001;
static const int return_home =      0b0010;
static const int entrymode_set =    0b0100;
static const int display_control =  0b1000;
static const int cursor_shift =     0b00010000;
static const int function_set =     0b00100000;
static const int set_cgram_addr =   0b01000000;
static const int set_ddram_addr =   0b10000000;

// start address first and second line, offset to set_ddram_addr
static const int first_line =       0;
static const int second_line =      0x40;

// flags for display entry mode
static const int entry_shiftinc =   0b0001;
static const int entry_left =       0b0010;

// flags for display and cursor control
static const int blink_on =         0b0001;
static const int cursor_on =        0b0010;
static const int display_on =       0b0100;

// flags for display and cursor shift
static const int move_right =       0b0100;
static const int display_move =     0b1000;

// flags for function set
static const int lcd_5x10dots =     0x04;
static const int lcd_2line =        0x08;
static const int lcd_8bitmode =     0x10;

// flag for backlight control
static const int backlight =        0b1000;


static const int lcd_enable_bit =   0b0100;

// By default these LCD display drivers are on bus address 0x27
static const int addr = 0x27;

// Modes for lcd_send_byte
static const int character = 1;
static const int command = 0;

/* Quick helper function for single byte transfers */
void i2c_write_byte(uint8_t val) {
//#ifdef i2c_default
    i2c_write_blocking(i2c_default, addr, &val, 1, false);
//#endif
}

static void lcd_toggle_enable(uint8_t val) {
    // Toggle enable pin on LCD display
    // We cannot do this too quickly or things don't work
#define DELAY_US 600
    sleep_us(DELAY_US);
    i2c_write_byte(val | lcd_enable_bit);
    sleep_us(DELAY_US);
    i2c_write_byte(val & ~lcd_enable_bit);
    sleep_us(DELAY_US);
}

// The display is sent a byte as two separate nibble transfers
void lcd_send_byte(uint8_t val, int mode) {
    uint8_t high = mode | (val & 0xF0) | backlight;
    uint8_t low = mode | ((val << 4) & 0xF0) | backlight;

    i2c_write_byte(high);
    lcd_toggle_enable(high);
    i2c_write_byte(low);
    lcd_toggle_enable(low);
}

void lcd_clear(void) {
    lcd_send_byte(clear_display, command);
}

// go to location on LCD
void lcd_set_cursor(int line, int position) {
    int val = set_ddram_addr + (!line ? first_line : second_line) + position;
    lcd_send_byte(val, command);
}

static inline void lcd_char(char val) {
    lcd_send_byte(val, character);
}

void lcd_string(const char *s) {
    while (*s) {
        lcd_char(*s++);
    }
}

void lcd_init() {
    gpio_set_function(PICO_DEFAULT_I2C_SDA_PIN, GPIO_FUNC_I2C);
    gpio_set_function(PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C);
    gpio_pull_up(PICO_DEFAULT_I2C_SDA_PIN);
    gpio_pull_up(PICO_DEFAULT_I2C_SCL_PIN);
    // Make the I2C pins available to picotool
    //bi_decl(bi_2pins_with_func(PICO_DEFAULT_I2C_SDA_PIN, PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C));

    lcd_send_byte(0x03, command);
    lcd_send_byte(0x03, command);
    lcd_send_byte(0x03, command);
    lcd_send_byte(0x02, command);

    lcd_send_byte(entrymode_set | entry_left, command);
    lcd_send_byte(function_set | lcd_2line, command);
    lcd_send_byte(display_control | display_on, command);
    lcd_clear();
}